exp/addons/api/v1alpha3/clusterresourcesetbinding_types.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ANCHOR: ResourceBinding
+
+// ResourceBinding keeps status of a resource that is applied to the cluster that ClusterResourceSetBinding belongs to.
+type ResourceBinding struct {
+	// Hash is the hash of a resource's data. This can be used to decide if a resource is changed.
+	// For "ApplyOnce" ClusterResourceSet.spec.strategy, this is no-op as that strategy does not act on change.
+	Hash string `json:"hash,omitempty"`
+	// LastAppliedTime identifies when this resource was last applied to the cluster.
+	// +optional
+	LastAppliedTime *metav1.Time `json:"lastAppliedTime,omitempty"`
+	// Applied is to track if a resource is applied to the cluster or not.
+	Applied bool `json:"applied,omitempty"`
+}
+
+// ANCHOR_END: ResourceBinding
+
+// ResourcesSetBinding keeps info on all of the resources in a ClusterResourceSet.
+type ResourcesSetBinding struct {
+	// Resources is a map of Secrets/ConfigMaps and their ResourceBinding.
+	// The map's key is a concatenated string of form: <resource-type>/<resource-name>.
+	Resources map[string]ResourceBinding `json:"resources,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterresourcesetbindings,scope=Namespaced,categories=cluster-api
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+
+// ClusterResourceSetBinding lists all matching ClusterResourceSets with the cluster it belongs to.
+type ClusterResourceSetBinding struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// ClusterResourceSetMap is a map of ClusterResourceSet name and its resources which is also a map.
+	ClusterResourceSetMap map[string]ResourcesSetBinding `json:"clusterresourcesetmap,omitempty"`
```

Comment:
These maps are also going to get a little complicated, it seems the final structure would be a map to a map to a struct?

Reply:
For a cluster, here we list all the ClusterResourceSets that are applied to that cluster along with the statuses of all the resources in those ClusterResourceSets.

I didn't put them under spec because these are not the desired state. 
```
apiVersion: clusterresourceset.cluster.x-k8s.io/v1alpha3
clusterresourcesetmap:
 clusterResourceSet1:
    resources:
      Secret/secret1:
        hash: _IcfqGpJWlhS5Az06XX4A6BXcTA=
        lastAppliedTime: "2020-05-28T08:51:06Z"
        successful: true
  clusterResourceSet1:
    resources:
      Configmap/cm1:
        hash: _IcfqGpJWlhS5Az06XX4A6BXcTA=
        lastAppliedTime: "2020-05-28T08:51:06Z"
        successful: false
```

Reply:
TIL I didn't know this was valid from an API perspective, it does break with things like patch helper though. Although I do agree it's not technically a spec, but more something that should be in a status, I don't think we can escape the fact that status would be lost in case of pivot/move.

Reply:
Yes, shouldn't go into status. I will move it under Spec then.

Reply:
Let's think a little more about the data model, having maps goes against API conventions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Reply:
Thanks for pointing out, I didn't know that. In that case, we will need to iterate all resources to to figure out if a resource has already been applied to a cluster or not.

Reply:
The underlying reason of not preferring maps is that the keys are unknown to the users in a map. Since CRSBinding is auto-generated (not generated/modified by users) and it will only be consumed by a controller, based on an offline chat with @ncdc, we are thinking maps may be acceptable for this scenario.

Reply:
Follow up: https://github.com/kubernetes-sigs/cluster-api/issues/3297